File: 662-maximum-width-of-binary-tree/maximum-width-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    using ui = unsigned int;
public:
    int widthOfBinaryTree(TreeNode* root) {
        vector<vector<ui>> widths;

        function<void(TreeNode*, int, ui)> dfs = [&] (TreeNode* node, int x, ui id) {
            if (!node) {
                return ;
            }

            if (widths.size() == x) {
                widths.push_back({id, id});
            } else {
                widths[x][0] = min<ui>(widths[x][0], id);
                widths[x][1] = max<ui>(widths[x][1], id);
            }

            dfs(node -> left, x + 1, 2 * id + 1);
            dfs(node -> right, x + 1, 2 * id + 2);
        };

        dfs(root, 0, (ui)0);
        int maxWidth = 0;
        for (vector<ui>& width: widths) {
            maxWidth = max<ui>(maxWidth, width[1] - width[0] + 1);
        }

        return maxWidth;
    }
};

File: 662-maximum-width-of-binary-tree/maximum-width-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int widthOfBinaryTree(TreeNode root) {
        class Box {
            public TreeNode node;
            public int id;
            public Box(TreeNode _node, int _id) {
                node = _node;
                id = _id;
            }
        }

        int maxWidth = 0;
        Deque<Box> q = new LinkedList<>();
        q.offerLast(new Box(root, 0));
        while (!q.isEmpty()) {
            maxWidth = Math.max(maxWidth, (q.peekLast().id) - (q.peekFirst().id) + 1);
            for (int ls = q.size(); ls > 0; ls--) {
                Box curr = q.pollFirst();
                if (curr.node.left != null) {
                    q.offerLast(new Box(curr.node.left, 2 * curr.id + 1));
                }
                if (curr.node.right != null) {
                    q.offerLast(new Box(curr.node.right, 2 * curr.id + 2));
                }
            }
        }

        return maxWidth;
    }
}